### ISSUE TITLE
Work parallelly using git?

### ISSUE THREAD

**Issue by CisBetter** [adventuregamestudio/ags#463]
# Is there a way to work parallelly using git?

For a small project of mine several developers are required to work parallelly.

They are each developing a separate room. Every day their changes are merged together using git.
For room scripts and room files this is not much of a problem.

---

**However**, if they add new sprites to the global sprite index, they overlap as their sprite numbers all start at the same number.
This makes merging their branches **almost impossible**.

Is there a way to do this (for example by manually changing the sprite numbers), or do I have to add all sprites created to the master branch myself?

**Comment by ivan-mogilko** [comment]
AGS games may only be merged partially. Sprites are all compressed to a single binary file, and so there is no way to merge different versions of it.
On the other hand you could supply original sprites as files in a subfolder(s) inside project's root: that way the editor will remember relative paths and will be able to recreate sprite file from these (I think you may even put acsprset.spr to gitignore.

As for the indexes, you can manually asign index to a sprite, this is a common practice, e.g. when you want to make a sequence of sprites to access using Random command or some formula. To do so select sprite, call context menu and choose "Change sprite number".
There are two caveats:
- Editor does not allow you to change it if it's assigned anywhere.
- It displays a warning *every time* you do that.
Neither makes sense to me (if it can find where sprite is used, it could change numbers on its own), so perhaps these are good targets for a fix request.

Unfortunately it does not automatically parse all rooms for the sprite usage, and cannot find it in script, so it's up to you to keep numbers consistent.

**Comment by CisBetter** [comment]
> and will be able to recreate sprite file

This is not working for me.
As you said I first took the SCI AGS Template (just as an example) and added every single sprite as a resource in a folder called resources/ in the projects root directory.
I replaced every existing sprite with the corresponding file in resource/
Next, I added acsprset.spr to gitignore.
AGS (using AGS Editor .NET (Build 3.4.2.1) ** BETA VERSION ** ) refused to open the project:

![screenshot from 2018-07-03 23-43-21](https://user-images.githubusercontent.com/17952058/42245911-eb9ede68-7f1a-11e8-8408-83b4ab701995.png)

Note: My resources/ folder contains all sub-folders of [the original resource folder](https://github.com/caesarcub/AGS-SCI-Template/tree/master/resources) in caesarcub's repository (plus the blue cup (spr000000.png) added to resources/).
Some of these sub-folders have spaces in their name.

Second Note: As I added acsprset.spr to gitignore before pushing to master for the first time, I do not have a copy of this file in my project root.

**Comment by ivan-mogilko** [comment]
Sorry, I could have been wrong about it. I believed such opportunity existed, but probably confused with something else.

**Comment by ivan-mogilko** [comment]
@CisBetter , thank you.
Right now I am planning to make certain changes to sprite management that would let AGS to fully restore sprite cache from the original sprites (assuming they are available, of course).

> Other binary files like audio might require different methods. 
Unlike sprites, audio files are already stored separately, AudioCache may be fully recreated if referenced original audio files are present.

The biggest problem is rooms, which need complete format change to be mergeable.

**Comment by jeancallisti** [comment]
This tutrial would be better off on the forums or on the wiki or directly in the help file. I don't even know what miracle made me stumble upon it here. :)

**Comment by CisBetter** [comment]
@jeancallisti I agree it might be hard for people not involved in development or the AGS issue tracker to find this tutorial. However, there are still a few things I'd like to update before publishing it somewhere else. I want to be sure this tutorial covers (almost) all things that could go wrong to prevent new users from getting frustrated by missing a command I don't find noteworthy.

Another important reason not to publish this article yet was mentioned by @ivan-mogilko.

> Right now I am planning to make certain changes to sprite management that would let AGS to fully  restore sprite cache from the original sprites (assuming they are available, of course).

Many things in this tutorial might be obsolete once he pushes his changes. And once the entire editor supports parallel development (*version 4.0? :man_shrugging:*) a few notes on what to look out for might be enough.

@ivan-mogilko
Would it make any sense to add **parallel development** and/or a **Git integration** to AGS? If so, we should create some new issues and assign them to the AGS 4 milestone // the AGS 4 label.

**Comment by ivan-mogilko** [comment]
> Another important reason not to publish this article yet was mentioned by @ivan-mogilko.

>> Right now I am planning to make certain changes to sprite management that would let AGS to fully restore sprite cache from the original sprites (assuming they are available, of course).

I cannot event tell when these changes will be ready and new version officially released; also users tend to use old versions for some times too, so that may not be a good idea to wait for that. Even then, you may divide instructions in sections specific for some versions.

> Would it make any sense to add parallel development and/or a Git integration to AGS? If so, we should create some new issues and assign them to the AGS 4 milestone // the AGS 4 label.

I do not think you need to ask me about whether it may or not make sense, surely if you believe that would make sense go ahead and make a feature suggestion.

AGS Editor has some source control integration API, but I never investigated how it works. Probably it relies on MSSCCI, but there also seem to be a plugin API for that.

**Comment by ivan-mogilko** [comment]
I believe this ticket has to be redone into a kind of a master list of problems of team work with AGS projects.

There were few changes since it was opened, firstly the Editor now supports recreating spritefile from sources, which makes it unnecessary to store the compiled spritefile under source control, you can just store the source image files (#1404).

There's a planned open sprite format still in works, but the progress seem slow (#1281).

It looks like the major issue which is mentioned in this ticket, but still not addressed, is the problem of conflicting sprite indexes. This is serious, and won't be resolved by #1281 either. I suppose it deserves its own ticket and a good discussion.

I [mentioned this in another ticket](https://github.com/adventuregamestudio/ags/issues/1281#issuecomment-1639502444), there are couple of ideas that i have in regards to this:
1. Implement a workspace setting that defines a range of sprite IDs to use when importing new sprites. Then each team member could define different range for themselves. This requires a new settings page in the project tree, which is only saved per-user, not in Game.agf.
2. Implement Sprite Folder configuration, where you may define the range of sprite IDs that imported sprites receive when you import in particular folder.
